### PR TITLE
fix: remove tasks from queue with 400 http response

### PR DESCRIPTION
**Pull request by Shahroz16**
closes: https://github.com/customerio/issues/issues/9443

Complete each step to get your pull request merged in. [Learn more about the workflow this project uses](https://github.com/customerio/customerio-android/develop/docs/dev-notes/GIT-WORKFLOW.md).
- [ ] [Assign members of your team](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/requesting-a-pull-request-review) to review the pull request.
- [ ] Wait for pull request status checks to complete. If there are problems, fix them until you see that [all status checks are passing](https://external-content.duckduckgo.com/iu/?u=https%3A%2F%2Fsymfony.com%2Fdoc%2F4.3%2F_images%2Fdocs-pull-request-symfonycloud.png&f=1&nofb=1).
- [ ] Wait until the pull request has been reviewed *and approved* by a teammate
- [ ] After pull request is approved, and you determine it's ready **add the label "Ready to merge"** to the pull request and it will automatically be merged.

### CODE REVIEW

File: sdk/src/sharedTest/java/io/customer/sdk/queue/QueueStorageIntegrationTest.kt [hunk 1]
```diff
@@ -109,6 +109,19 @@ class QueueStorageIntegrationTest : BaseTest() {
...
+    fun givenDeleteGroupTask_expectUpdateInventory() {
+        val givenStartOfTheGroup = QueueTaskGroup.IdentifyProfile(String.random)
+
+        queueStorage.create(String.random, String.random, givenStartOfTheGroup, null)
+        queueStorage.create(String.random, String.random, null, listOf(givenStartOfTheGroup))
+        queueStorage.create(String.random, String.random, null, listOf(givenStartOfTheGroup))
+
+        val itemsDeleted = queueStorage.deleteGroup(givenStartOfTheGroup.toString())
+        itemsDeleted.count() shouldBeEqualTo 3
+        queueStorage.getInventory().count() shouldBeEqualTo 0
```
Comment (mrehan27) [inline sdk/src/sharedTest/java/io/customer/sdk/queue/QueueStorageIntegrationTest.kt:122]: Shouldn't we also need to add test for case when there are other tasks in the queue as well to make sure only the desired tasks have been deleted?

**Review by mrehan27 — CHANGES_REQUESTED** [review]

File: sdk/src/sharedTest/java/io/customer/sdk/queue/QueueStorageIntegrationTest.kt (hunk 1 above)
Comment (Shahroz16) [reply sdk/src/sharedTest/java/io/customer/sdk/queue/QueueStorageIntegrationTest.kt:122]: I added those tests, but forgot to push them 😆 
Screenshot 2023-03-02 at 9 31 32 PM

File: sdk/src/main/java/io/customer/sdk/queue/QueueStorage.kt [hunk 2]
```diff
@@ -104,6 +105,20 @@ internal class QueueStorageImpl internal constructor(
...
+
+        val groupStart = inventory.filter { it.groupStart == groupStartTask }
+        val groupMember = inventory.filter { task ->
+            task.groupMember != null && groupStartTask in task.groupMember
+        }
+
+        val tasksToBeDeleted = groupStart + groupMember
+        tasksToBeDeleted.forEach { delete(it.taskPersistedId) }
+
+        return tasksToBeDeleted
```
[the remark this answers is not in the document]
Comment (Shahroz16) [reply sdk/src/main/java/io/customer/sdk/queue/QueueStorage.kt:119]: I could add some comments but this [doc](https://github.com/customerio/customerio-ios/blob/main/docs/dev-notes/BACKGROUND-QUEUE.md) could help add more clarity on the `groups` concept.

[the remark this answers is not in the document]
Comment (mrehan27) [reply sdk/src/main/java/io/customer/sdk/queue/QueueStorage.kt:122]: Thanks for sharing the docs. Also, the comments you now added are enough to read the code 😄

**Review by mrehan27 — APPROVED** [review]
I don't have a very good understanding of queue implementation yet 😅 But tried going through everything and looked good to me 👍🏻

File: sdk/src/main/java/io/customer/sdk/queue/QueueRunRequest.kt [hunk 3]
```diff
@@ -60,25 +64,47 @@ internal class QueueRunRequestImpl internal constructor(
...
+                        logger.debug("queue deleting task $nextTaskStorageId because it will always fail")
+
+                        // check if its a groupStart task, if so delete the group otherwise just delete the task
+                        if (nextTaskToRunInventoryItem.groupStart == null) {
+                            queueStorage.delete(nextTaskStorageId)
+                        } else {
+                            queueStorage.deleteGroup(groupStartTask = nextTaskToRunInventoryItem.groupStart)
+                        }
+
+                        // since failed task isn't being saved, and the group is also deleted no need to update `lastFailedTask`
```
Comment (levibostian) [inline sdk/src/main/java/io/customer/sdk/queue/QueueRunRequest.kt:86]: 🤔 I understand that there is no need to updating `lastFailedTask`, but is there a problem with updating `lastFailedTask`? 

I am worried about a side-effect happening from this code decision in the future. What if the logic of `goToNextTask` changes and passing `null` here as `lastFailedTask` causes a bug? 

I am wondering if maybe it's a better idea for us to use a non-null value here even if it doesn't make any difference to the code today. Maybe it will make a difference in the future? The task did indeed fail, so makes sense to use a non-null value?

File: sdk/src/main/java/io/customer/sdk/api/HttpRequestRunner.kt [hunk 4]
```diff
@@ -79,19 +82,24 @@ internal class HttpRequestRunnerImpl(
...
+                        errorBodyString // if we can't parse the error body json, the raw response string is good to capture.
 
                     parseCustomerIOErrorBody(errorBodyString)?.message?.let { parsedErrorMessage ->
                         errorMessage = parsedErrorMessage
                     }
                 }
 
-                val customerIOError = CustomerIOError.UnsuccessfulStatusCode(statusCode, errorMessage)
+                val customerIOError =
+                    CustomerIOError.UnsuccessfulStatusCode(statusCode, errorMessage)
```
Comment (levibostian) [inline sdk/src/main/java/io/customer/sdk/api/HttpRequestRunner.kt:102]: Not a blocker to this PR but I am curious. 

What is causing these format changes? ktlint? A Intellij plugin? 

It would be ideal if the formatting tools that were used in the project were the same across all engineers on the team. So if, for example, this change was caused by an intellij plugin, I think that plugin should be disabled or that rule gets applied to ktlint (or another tool that the team uses).

File: sdk/src/main/java/io/customer/sdk/queue/QueueStorage.kt [hunk 5]
```diff
@@ -104,6 +105,23 @@ internal class QueueStorageImpl internal constructor(
         return jsonAdapter.fromJsonOrNull(fileContents)
     }
 
+    override fun deleteGroup(groupStartTask: String): List<QueueTaskMetadata> {
+        val inventory = getInventory()
+
+        // find the start of the group.
+        val groupStart = inventory.filter { it.groupStart == groupStartTask }
```
Comment (levibostian) [inline sdk/src/main/java/io/customer/sdk/queue/QueueStorage.kt:112]: I might have found an edge case that could cause a bug. 

If the caller of `deleteGroup` ever calls `delete(taskStorageId:)` before calling `deleteGroup`, this might cause a bug since the task `groupStart` will no longer exist. 

I think it would be good for this function to handle that edge case and this could be another reason for encapsulating the delete task + delete group members into this class could be a good idea.

File: sdk/src/sharedTest/java/io/customer/sdk/api/HttpRequestRunnerTest.kt [hunk 6]
```diff
@@ -192,6 +192,18 @@ class HttpRequestRunnerTest : BaseTest() {
         actual shouldBeEqualTo expected
     }
 
+    @Test
+    fun performHttpRequest_given400_expectNotPausedHttpRequestsAndReturnFailure(): Unit = runBlocking {
```
Comment (levibostian) [inline sdk/src/sharedTest/java/io/customer/sdk/api/HttpRequestRunnerTest.kt:196]: Good test to add. 

Does this test exist for the iOS SDK PR? I think it's a valuable test.

File: sdk/src/sharedTest/java/io/customer/sdk/queue/QueueRunRequestTest.kt [hunk 7]
```diff
@@ -67,41 +74,109 @@ class QueueRunRequestTest : BaseTest() {
     }
 
     @Test
-    fun test_start_givenHttpRequestsPaused_expectDontDeleteTask_expectDontUpdateTask(): Unit = runBlocking {
+    fun test_run_givenRunTask400Failure_expectDeleteTask(): Unit = runBlocking {
```
Comment (levibostian) [inline sdk/src/sharedTest/java/io/customer/sdk/queue/QueueRunRequestTest.kt:77]: Do we have integration tests setup for the Queue? Aka: tests that do not require so much mocking such as `storageMock`? 

I think the tests you added are testing good edge cases but they could be higher quality and less annoying to maintain (since we don't use as many mocks) if they were integration tests.

File: sdk/src/sharedTest/java/io/customer/sdk/queue/QueueRunRequestTest.kt [hunk 8]
```diff
@@ -67,41 +74,109 @@ class QueueRunRequestTest : BaseTest() {
...
+            QueueModifyResult(
+                true,
+                QueueStatus(siteId, 0)
+            )
+        )
         runRequest.run()
 
-        verify(storageMock, never()).delete(anyOrNull())
-        verify(storageMock, never()).update(anyOrNull(), anyOrNull())
+        verify(storageMock).delete(givenTaskId)
```
Comment (levibostian) [inline sdk/src/sharedTest/java/io/customer/sdk/queue/QueueRunRequestTest.kt:98]: Would be good to check that `storage.deleteGroup` was *not* called.

File: sdk/src/sharedTest/java/io/customer/sdk/queue/QueueRunRequestTest.kt [hunk 9]
```diff
@@ -67,41 +74,109 @@ class QueueRunRequestTest : BaseTest() {
...
-        whenever(storageMock.delete(any())).thenReturn(QueueModifyResult(true, QueueStatus(siteId, 0)))
+        whenever(storageMock.deleteGroup(eq(givenTaskId))).thenReturn(givenListOfQueueTaskMetadata)
 
         runRequest.run()
 
-        verify(storageMock).delete(givenTaskId)
-        verify(storageMock).delete(givenTaskId2)
-        verify(runnerMock).runTask(givenQueueTask)
-        verify(runnerMock).runTask(givenQueueTask2)
+        verify(storageMock).deleteGroup(givenGroupOfTasks.toString())
```
Comment (levibostian) [inline sdk/src/sharedTest/java/io/customer/sdk/queue/QueueRunRequestTest.kt:124]: Would be good to check that `storage.delete` was *not* called.

File: sdk/src/sharedTest/java/io/customer/sdk/queue/QueueStorageIntegrationTest.kt [hunk 10]
```diff
@@ -109,6 +109,49 @@ class QueueStorageIntegrationTest : BaseTest() {
         queueStorage.getInventory().count() shouldBeEqualTo 0
     }
 
+    @Test
+    fun givenDeleteGroupTask_expectAllTasksInGroupToDelete() {
+        val givenStartOfTheGroup = QueueTaskGroup.IdentifyProfile(String.random)
+
+        queueStorage.create(String.random, String.random, givenStartOfTheGroup, null)
+        queueStorage.create(String.random, String.random, null, listOf(givenStartOfTheGroup))
+        queueStorage.create(String.random, String.random, null, listOf(givenStartOfTheGroup))
```
Comment (levibostian) [inline sdk/src/sharedTest/java/io/customer/sdk/queue/QueueStorageIntegrationTest.kt:118]: What about this edge case: 

What if we had this:
```
```
```suggestion
        queueStorage.create(String.random, String.random, groupStartOfAnotherGroup, listOf(givenStartOfTheGroup))
        queueStorage.create(String.random, String.random, null, listOf(givenStartOfTheGroup))
```

So a task was the member of a group *and* was also the start of a separate group. 

If *that* task gets deleted (because it's a member of a group), should we also delete all of the tasks that *it* starts? 

I think that register push token is one of these edge cases. It belongs to the identify profile group but it also starts it's own group. Delete device token is a member of that group that register starts.

File: sdk/src/sharedTest/java/io/customer/sdk/queue/QueueStorageIntegrationTest.kt [hunk 11]
```diff
@@ -109,6 +109,49 @@ class QueueStorageIntegrationTest : BaseTest() {
...
+        val givenStartOfAnotherGroup = QueueTaskGroup.RegisterPushToken(String.random)
+
+        queueStorage.create(String.random, String.random, givenStartOfTheGroup, null)
+        queueStorage.create(String.random, String.random, null, listOf(givenStartOfTheGroup))
+        queueStorage.create(String.random, String.random, null, null)
+        queueStorage.create(String.random, String.random, null, listOf(givenStartOfAnotherGroup))
+
+        val itemsDeleted = queueStorage.deleteGroup(givenStartOfTheGroup.toString())
+        itemsDeleted.count() shouldBeEqualTo 2
+        queueStorage.getInventory().count() shouldBeEqualTo 2
```
Comment (levibostian) [inline sdk/src/sharedTest/java/io/customer/sdk/queue/QueueStorageIntegrationTest.kt:137]: Be good to check that the 2 tasks in inventory are tasks you expected to be in the inventory.

File: sdk/src/sharedTest/java/io/customer/sdk/queue/QueueIntegrationTests.kt [hunk 12]
```diff
@@ -27,24 +29,85 @@ class QueueIntegrationTests : BaseTest() {
...
+        mockWebServer.enqueue(MockResponse().setResponseCode(400))
+
+        queue.queueTrack(givenIdentifier, String.random, EventType.event, emptyMap())
+        mockWebServer.enqueue(MockResponse().setResponseCode(404))
+
         queue.run()
 
-        // expect all of tasks to run and run successfully
-        queueStorage.getInventory().count() shouldBeEqualTo 0
+        // expect tasks with 400 still present
```
Comment (levibostian) [inline sdk/src/sharedTest/java/io/customer/sdk/queue/QueueIntegrationTests.kt:84]:
```suggestion
        // expect tasks that did not receive a 400 response still present in queue
```

File: sdk/src/sharedTest/java/io/customer/sdk/queue/QueueIntegrationTests.kt [hunk 13]
```diff
@@ -27,24 +29,85 @@ class QueueIntegrationTests : BaseTest() {
...
+        queueStorage.getInventory().count() shouldBeEqualTo 1
```
Comment (levibostian) [inline sdk/src/sharedTest/java/io/customer/sdk/queue/QueueIntegrationTests.kt:85]: I think it would be good to check that the 1 item remaining in the queue is the item we expect to see (the 404 response). 

You have done a lot of work on this PR. I'll help with this task.

File: sdk/src/sharedTest/java/io/customer/sdk/queue/QueueIntegrationTests.kt [hunk 14]
```diff
@@ -27,24 +29,85 @@ class QueueIntegrationTests : BaseTest() {
...
         mockWebServer.requestCount shouldBeEqualTo 3
     }
+
+    @Test
+    fun givenRunQueueAndMultipleTaskGroupsFailWith400Tasks_expectQueueToBeEmpty(): Unit =
```
Comment (levibostian) [inline sdk/src/sharedTest/java/io/customer/sdk/queue/QueueIntegrationTests.kt:90]: Let's also make sure to test the edge case of a task being deleted because it's a member of a group but it's also the start of a group. 

I'll help add this test to keep the PR moving along.

File: sdk/src/sharedTest/java/io/customer/sdk/queue/QueueStorageIntegrationTest.kt [hunk 15]
```diff
@@ -132,9 +134,74 @@ class QueueStorageIntegrationTest : BaseTest() {
...
+        queueStorage.getInventory().count() shouldBeEqualTo 0
+    }
+
+    @Test
+    fun givenDeleteGroupTask_givenMembersTasksBelongToDifferentGroups_givenDifferentOrder_expectAllStartTasksAndTheirMembersToBeDeleted() {
+        val givenStartOfTheGroup = QueueTaskGroup.IdentifyProfile(String.random)
+        val givenStartOfAnotherGroup = QueueTaskGroup.RegisterPushToken(String.random)
+
+        queueStorage.create(String.random, String.random, null, listOf(givenStartOfTheGroup))
+        queueStorage.create(String.random, String.random, givenStartOfTheGroup, null)
```
Comment (levibostian) [inline sdk/src/sharedTest/java/io/customer/sdk/queue/QueueStorageIntegrationTest.kt:176]: This scenario *should* never occur where a BQ task that is the start of a group exists *after* a BQ task that belongs to that group. 

So, I don't see this test function adding value. Do you see a reason to keep this test function?

File: sdk/src/sharedTest/java/io/customer/sdk/queue/QueueStorageIntegrationTest.kt [hunk 16]
```diff
@@ -132,9 +134,74 @@ class QueueStorageIntegrationTest : BaseTest() {
...
+            listOf(givenStartOfTheGroup)
+        )
+
+        val itemsDeleted = queueStorage.deleteGroup(givenStartOfTheGroup.toString())
+        itemsDeleted.count() shouldBeEqualTo 4
+        queueStorage.getInventory().count() shouldBeEqualTo 1
+    }
+
+    @Test
+    fun givenDeleteGroupTask_givenIncorrectQueue_expectNotToGetInfiniteLoop() {
```
Comment (levibostian) [inline sdk/src/sharedTest/java/io/customer/sdk/queue/QueueStorageIntegrationTest.kt:192]:
```suggestion
    fun givenDeleteGroupTask_givenTaskStartsAndBelongsToSameGroup_expectNotToGetInfiniteLoop() {
```

I think defining "incorrect queue" would document what this test function is ultimately trying to test.

File: sdk/src/sharedTest/java/io/customer/sdk/queue/QueueStorageIntegrationTest.kt (hunk 15 above)
Comment (Shahroz16) [reply sdk/src/sharedTest/java/io/customer/sdk/queue/QueueStorageIntegrationTest.kt:176]: It's still an edge case that I like to keep, to make sure we safe guard against it.

**Review by levibostian — APPROVED** [review]
